src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+  @arg(flag="ds", doc="Inserted into the read group header") val description: Option[String] = None,
+  @arg(flag="dt", doc="Date the run was produced, to insert into the read group header") val runDate: Option[Iso8601Date] = None,
+  @arg(flag="so", doc="The sort order for the output sam/bam file.") val sortOrder: SortOrder = SortOrder.queryname,
+  @arg(doc="Minimum base quality allowed in the input fastq.  An exception will be thrown if a quality is less than this value.") val minQ: Int = 0,
+  @arg(doc="Maximum base quality allowed in the input fastq.  An exception will be thrown if a quality is greater than this value.") val maxQ: Int = SAMUtils.MAX_PHRED_SCORE,
+  @arg(doc="Allow (and ignore) empty lines") val allowAndIgnoreEmptyLines: Boolean = false,
+  @arg(doc="Maximum mismatches for a barcode to be considered a match.") val maxMismatches: Int = 1,
+  @arg(doc="Minimum difference between number of mismatches in the best and second best barcodes for a barcode to be considered a match.") val minMismatchDelta: Int = 2,
+  @arg(doc="Maximum allowable number of no-calls in a barcode read before it is considered unmatchable.") val maxNoCalls: Int = 2,
+  @arg(doc="Also match against the reverse complement of the sample barcodes (in-line sample barcodes only).") val matchReverseComplement: Boolean = false
+) extends FgBioTool with LazyLogging {
+
+  import DemuxFastqs._
+
+  // Check all input and output files are readable or writable respectively.
+  Io.assertReadable(fastq1 ++ fastq2 ++ fastq7 ++ fastq5)
+  Io.assertReadable(this.sampleSheet)
+  Io.assertWritableDirectory(this.output)
+  Io.assertCanWriteFile(this.metrics)
+
+  override def execute(): Unit = {
+    // Make some data structures to help us on our way
+    val r1ReadStructure          = ReadStructure(rs1)
+    val r2ReadStructure          = rs2.map(ReadStructure(_))
+    val i7ReadStructure          = rs7.map(ReadStructure(_))
+    val i5ReadStructure          = rs5.map(ReadStructure(_))
+
+    // Verify that the template bases are only on read one or read two
+    if (r1ReadStructure.template.isEmpty)           throw new IllegalArgumentException("No template bases found in the read structure for read one.")
+    if (r2ReadStructure.exists(_.template.isEmpty)) throw new IllegalArgumentException("No template bases found in the read structure for read two.")
+    if (i7ReadStructure.exists(_.template.nonEmpty)) throw new IllegalArgumentException("Template bases not allowed in the i7 read.")
+    if (i5ReadStructure.exists(_.template.nonEmpty)) throw new IllegalArgumentException("Template bases not allowed in the i5 read.")
+
+    val fastqs = Seq(fastq1, fastq2, fastq7, fastq5)
+    val readStructures: Seq[ReadStructure] = Seq(Some(r1ReadStructure), r2ReadStructure, i7ReadStructure, i5ReadStructure).flatten
+    if (readStructures.length != fastqs.count(_.nonEmpty)) throw new IllegalArgumentException("The # of read structures did not match the number of fastqs.")
```

Comment:
This check shoud probably zip the fastqs with the readStructure and assert `f.isEmpty == rs.isEmpty` otherwise you will do odd things if a user messes up and provides the same number but different fastqs and read structures.